Add a tag-order-preserving option for `Parser.get_annotation`

woodrush opened the pull request:
In the raw XML in the Manga109 dataset, the bounding box data in the `page` tag is not sorted by its annotation type, and each bounding box type appears in an arbitrary order. Since the output of the `get_annotation` function is sorted by the bounding box type, the order of the tags in the original XML file is destroyed in this output.

This is particularly troublesome when managing the raw form of the XML, especially when trying to use this library to edit and fix the dataset's annotations.

This pull request adds the `Parser.get_annotation_ordered` function, where the ordering of each bounding box tag in the raw XML data is preserved. Due to the order-preserving feature, the format of the output dictionary slightly differs from the `Parser.get_annotation` function. This is mentioned in the README as:

```
# Caution: Due to the aforementioned feature, the format of the output
# dictionary differs from the `get_annotation` function.
```

Also, it is stated in the README that the ordering of the tags do not carry any useful information about the contents of the data:

```
# Note that the ordering of
# the bounding box tags does not carry any useful information about the
# contents of the data.
```

I have also written a pytest, which I have checked that it is passing.

kktsubota commented on manga109api/manga109api.py:
```diff
@@ -17,6 +19,66 @@ def __init__(self, root_dir):
+                    d["@"+k] = d.pop(k)
+            return d
+
+        with (self.root_dir / annotation_type / (book + ".xml")).open("rt", encoding= 'utf-8') as f:
```
You should remove a space between `encoding=` and `'utf-8'`
I find `get_annotation` also has a space in the same place.
So, I will thank you if you remove it.

woodrush replied:
Thanks, fixed it in the new commit!

kktsubota reviewed (approved):
LGTM

matsui528 commented:
It would be better to maintain the order of tags. Can you replace the all original code?

ku21fan commented on manga109api/manga109api.py:
```diff
@@ -17,6 +19,66 @@ def __init__(self, root_dir):
+        def keys_renamed(d):
+            for k in list(d.keys()):
+                try:
+                    if str(d[k]) == str(int(d[k])):
```
Which key needs this if function? 
Could you add a simple comment (e.g. `# ~~~ key needs this`) here?

woodrush replied:
Thanks, in the new code, I added a clarification of the keys that are affected:
https://github.com/manga109/manga109api/blob/add-ordered-output-func/manga109api/manga109api.py#L47

ku21fan reviewed (approved):
LGTM!

ku21fan commented on manga109api/manga109api.py:
```diff
@@ -17,6 +19,66 @@ def __init__(self, root_dir):
+                try:
+                    if str(d[k]) == str(int(d[k])):
+                        d["@"+k] = int(d.pop(k))
+                    else:
+                        d["@"+k] = d.pop(k)
+                except ValueError as e:
+                    d["@"+k] = d.pop(k)
```
'else' case and 'except' case do same thing `d["@"+k] = d.pop(k)`
thus, we should remove one of them? (e.g. remove 'if' & 'else' and only use try/except)

woodrush replied:
I thought for a moment too for simplifying this part. The complicated issue is that this function here is actually trying to capture three patterns:

- (1): The `str(int(x))` throws an exception and fails, since `x` cannot be interpreted as an integer literal, such as when `x == "some text"` . In this case, return `x`.
- (2): The `str(int(x))` succeeds, but `if str(x) == str(int(x))` is False, such as when `x == "00001234"`. In this case, return `x`.
- (3): The `str(int(x))` succeeds, and `if str(x) == str(int(x))` is True. In this case, return `int(x)`.

In the new code here https://github.com/manga109/manga109api/blob/add-ordered-output-func/manga109api/manga109api.py#L50 , I chose to throw a `ValueError` in the exception in the (2) case. Since (1) also throws a ValueError (when `str(int(x))`, or more precisely, when `int(x)` fails), this can capture both cases. Case (3) is handled inside the exception.

One concern was whether if the `ValueError` exception is a natural choice to `throw` in the `except` block. What are your thoughts about this and the overall code?

ku21fan replied:
I see. I understand your purpose now :)

> One concern was whether if the ValueError exception is a natural choice to throw in the except block.

I feel slightly unnatural because (`if str(x) == str(int(x))` is False) is not ValueError?

In this case, in my opinion, the previous code was easy to understand. (sorry for bothering you)
I needed simple comments/examples to understand the purpose.
```
            try:
                if str(t) == str(int(t)):
                    return int(t)
                else:
                    return t    # for the case, such as t = "00001234"
            except ValueError as e:
                return t     # for the case, such as t = "some text" or t = "000012ab"
```

woodrush replied:
Thanks, yes, I was feeling that raising a ValueError was a little unnatural too. I'm glad the previous code was easy to understand, no problem at all, I'll put it back and also put some comments as in your suggestion :)

ku21fan commented on manga109api/manga109api.py:
```diff
@@ -17,22 +21,77 @@ def __init__(self, root_dir):
 
-    def get_annotation(self, book, annotation_type="annotations"):
+
+    def get_annotation(self, book, annotation_type="annotations", separate_by_tag=True):
```
Since we use `annotation_types = ["frame", "face", "body", "text"]` above, 
I guess we should change `annotation_type` to `annotation_dir` or something here.

And thanks to `separate_by_tag`, we can get the same result as before. Nice :D

woodrush replied:
> I guess we should change `annotation_type` to `annotation_dir` or something here.

Oh, sorry I didn't notice that, in fact I was overloading `annotation_type` in a `for` block... Since we have the option separate_by_"tag" , I guess I should change `annotation_types = ["frame", "face", "body", "text"]` to `annotation_tags = ["frame", "face", "body", "text"]`.

> And thanks to `separate_by_tag`, we can get the same result as before. Nice :D

Thanks! After writing some filtering code, I thought it was better in a lot of ways to have a mode to return the original format.

matsui528 commented on README.md:
```diff
@@ -127,6 +127,49 @@ pprint(annotation["page"][3])  # the data of the 4th page of "ARMS"
+# Caution: Due to the aforementioned feature, the format of the output
+# dictionary differs from the `get_annotation` function.
+
+# (4-1) Tag ordering in `page`
```
(4-1)?

woodrush replied:
After looking again at the readme, I thought this line was redundant, so I will be deleting it altogether.

matsui528 commented on manga109api/manga109api.py:
```diff
@@ -17,22 +21,77 @@ def __init__(self, root_dir):
+        def int_literals_to_int(t):
+            # Convert certain keys to integer types.
+            # The following keys will be affected with this function:
+            # '@index', '@width', '@height', '@xmax', '@ymax', '@xmin', '@ymin'
```
As @ku21fan suggests, can you write some comments and examples?

woodrush replied:
Thanks, added a more detailed description of the function.

matsui528 commented on manga109api/manga109api.py:
```diff
@@ -17,22 +21,77 @@ def __init__(self, root_dir):
+            else:
+                return int(t)
+
+        def rename_keys(d):
```
Can you show an example of input/output?

woodrush replied:
Added a docstring containing a description as well as an example.

woodrush commented:
Revised the description of the ordering feature in the readme according to the newest code (it is now an option of an existing function now, instead of a new function)

woodrush commented:
I squashed the two commits into one for simplicity.

Also, since I noticed that this pull request eliminates the need for the `xmltodict` library, I deleted the dependency from requirements.txt (which became empty now) and also deleted the import statement for `xmltodict`.

woodrush committed:
Add an order-preserving output option for the `Parser.get_annotation` function

woodrush commented:
I'm sorry for making a lot of revisions - I renamed `rename_keys` to `formatted_dict` since it does a little more than just renaming the keys, since it:
```
- Also applies `int_literals_to_int` to each value of the given dict.
```
I added this line to the function's description as well.

Also, this function is a lot more simpler now:
https://github.com/manga109/manga109api/pull/17/files#diff-691a21da143771320816a57c506cefe7R66

I also rebased this branch so that the branch is not behind by any commits compared to `master`.

matsui528 reviewed (approved):
LGTM
